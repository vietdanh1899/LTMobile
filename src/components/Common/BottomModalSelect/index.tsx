import React, { Component } from 'react';
import {
  Text,
  View,
  Dimensions,
  StyleSheet,
  Animated,
  TouchableOpacity,
  TouchableWithoutFeedback,
  ScrollView,
} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import _ from 'lodash';
import { Props as ListCheckBoxProps } from '@components/common/ListCheckBox';
import ListCheckBox from '../ListCheckBox';

const { width, height } = Dimensions.get('window');
const styles = StyleSheet.create({
  wrapper: {
    ...StyleSheet.absoluteFillObject,
    // height: 0,
  },
  overlayModal: {
    ...StyleSheet.absoluteFillObject,
    // backgroundColor: 'rgba(0,0,0,0.65)',
  },
  containerStyle: {
    flex: 1,
  },
});

export interface Props extends Pick<ListCheckBoxProps, 'checkBoxProps'> {
  iconColor?: string;
  closeIcon?: boolean;
  title?: string;
  titleConfirm?: string;
  titleStyles?: any;
  buttonConfirmStyles?: any;
  titleConfirmStyles?: any;
  data?: any;
  onChange?: any;
  onConfirm?: any;
  multi?: boolean;
  defaultVal?: Array<string | number>;
  wrapStyles?: any;
  headerContainerStyles?: any;
  renderContent?: any;
  contentContainerStyles?: any;
}

interface State {
  modalFilter: boolean;
  modalAnimatedValue: any;
  preValue: Array<string | number>;
  assistValue: Array<string | number>;
}

class BottomModalSelect extends Component<Props, State> {
  static defaultProps = {
    closeIcon: true,
    multi: false,
  };

  constructor(props: Props) {
    super(props);
    const { defaultVal, data } = this.props;
    this.state = {
      modalFilter: false,
      modalAnimatedValue: new Animated.Value(0),
      preValue: defaultVal || (data ? [data[0].id] : []),
      assistValue: defaultVal || (data ? [data[0].id] : []),
    };
  }

  handleOpenFilter = () => {
    const { modalFilter, modalAnimatedValue } = this.state;

    if (modalFilter) {
      return;
    }
    this.setState({ modalFilter: true }, () => {
      Animated.timing(modalAnimatedValue, {
        toValue: 1,
        duration: 200,
        useNativeDriver: true,
      }).start();
    });
  };

  handlePressDone = () => {
    const { modalAnimatedValue, preValue } = this.state;
    const { onConfirm } = this.props;
    // console.log('done', preValue);
    if (!_.isUndefined(onConfirm)) {
      onConfirm(preValue);
    }
    this.setState({
      assistValue: preValue,
    });

    Animated.timing(modalAnimatedValue, {
      toValue: 0,
      duration: 150,
      useNativeDriver: true,
    }).start(() => {
      this.setState({ modalFilter: false });
    });
  };

  handlePressClose = () => {
    const { modalAnimatedValue } = this.state;
    Animated.timing(modalAnimatedValue, {
      toValue: 0,
      duration: 150,
      useNativeDriver: true,
    }).start(() => {
      this.setState({ modalFilter: false });
    });
  };

  onSelect = (preValue: any) => {
    const { onChange, multi } = this.props;
    const preValueArrTemp = [];
    preValueArrTemp.push(preValue);
    // console.log('preValue', preValue);
    if (!_.isUndefined(onChange)) {
      onChange(preValue);
    }
    this.setState({
      preValue: multi ? preValue : preValueArrTemp,
    });
  };

  render() {
    const { modalFilter, modalAnimatedValue, assistValue } = this.state;
    const {
      closeIcon,
      title,
      titleConfirm,
      titleStyles,
      buttonConfirmStyles,
      titleConfirmStyles,
      data,
      multi,
      wrapStyles,
      headerContainerStyles,
      renderContent,
      contentContainerStyles,
      iconColor,
      checkBoxProps,
    } = this.props;

    const opacity = modalAnimatedValue;
    const translateY = modalAnimatedValue.interpolate({
      inputRange: [0, 1],
      outputRange: [400, 0],
    });
    return (
      <View
        //  style={styles.containerStyle}
        testID="BottomModalSelect"
      >
        {modalFilter ? (
          <View
            style={styles.wrapper}
            pointerEvents={modalFilter ? 'auto' : 'none'}
          >
            <TouchableWithoutFeedback onPress={this.handlePressDone}>
              <Animated.View style={[styles.overlayModal, { opacity }]} />
            </TouchableWithoutFeedback>
            <Animated.View
              style={{
                position: 'absolute',
                bottom: 0,
                left: 0,
                transform: [{ translateY }],
              }}
            >
              <View
                style={[
                  {
                    width,
                    backgroundColor: '#F1F8FE',
                    height: height / 2 - 50,
                    paddingHorizontal: 20,
                    paddingBottom: 30,
                  },
                  wrapStyles,
                ]}
              >
                <View
                  style={[
                    {
                      flexDirection: 'row',
                      paddingBottom: 20,
                      paddingTop: 10,
                      backgroundColor: '#F1F8FE',
                      alignItems: 'center',
                      justifyContent: 'space-between',
                      // borderWidth: 1,
                      // flex: 1,
                    },
                    headerContainerStyles,
                  ]}
                >
                  <TouchableOpacity
                    style={{
                      width: '10%',
                      justifyContent: 'center',
                      // borderWidth: 1,
                    }}
                    onPress={this.handlePressClose}
                  >
                    {closeIcon ? (
                      <Icon
                        name="close"
                        size={24}
                        color={iconColor || '#000000'}
                      />
                    ) : null}
                  </TouchableOpacity>
                  <View
                    style={{
                      width: '60%',
                      alignItems: 'center',
                      justifyContent: 'center',
                      // borderWidth: 1,
                    }}
                  >
                    <Text
                      style={[
                        {
                          fontSize: 24,
                          fontWeight: 'bold',
                          color: '#00358E',
                          ...titleStyles,
                        },
                      ]}
                    >
                      {title || 'Ch·ªçn'}
                    </Text>
                  </View>
                  <TouchableOpacity
                    style={{
                      backgroundColor: '#F1F8FE',
                      padding: 10,
                      width: '20%',
                      // borderWidth: 1,
                      alignItems: 'center',
                      ...buttonConfirmStyles,
                    }}
                    onPress={this.handlePressDone}
                  >
                    <Text style={{ color: '#315DF7', ...titleConfirmStyles }}>
                      {titleConfirm || 'Xong'}
                    </Text>
                  </TouchableOpacity>
                </View>
                <View
                  style={[
                    {
                      paddingHorizontal: 15,
                      paddingVertical: 20,
                      flex: 2,
                      backgroundColor: '#FFFFFF',
                      borderRadius: 10,
                      // borderWidth: 1,
                    },
                    contentContainerStyles,
                  ]}
                >
                  <ScrollView>
                    {renderContent || (
                      <ListCheckBox
                        data={data || []}
                        onChange={(value: any) => this.onSelect(value)}
                        single={!multi}
                        checkBoxProps={{
                          textStyle: { color: '#333333' },
                          ...checkBoxProps,
                        }}
                        assistVal={assistValue}
                      />
                    )}
                  </ScrollView>
                </View>
              </View>
            </Animated.View>
          </View>
        ) : null}
      </View>
    );
  }
}
export default BottomModalSelect;
