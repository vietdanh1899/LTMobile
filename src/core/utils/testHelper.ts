/* eslint-disable newline-per-chained-call */
/* eslint-disable no-restricted-syntax */
/* eslint-disable import/prefer-default-export */
import { shallow, ShallowWrapper } from 'enzyme';
import React, { ReactElement, Component } from 'react';
import toJson from 'enzyme-to-json';

export enum EnumComponent {
  SINGLE = 'single',
  HOC = 'hoc',
}

/**
 *
 * @param node React Native Component
 * @param listOfProps List of Props that you pass into current Component
 * @param type Type of current Component ('single' or 'hoc')
 */
export function getComponent<C extends Component, P = C['props'], S = C['state']>(
  node: ReactElement<P>,
  listOfProps: any,
  type: EnumComponent,
): ShallowWrapper<P, S, C> {
  const nodeWithProps = React.cloneElement(
    node,
    listOfProps,
  );
  let component: any;
  switch (type) {
    case EnumComponent.HOC:
      component = shallow(nodeWithProps).first().shallow().first().shallow().first().shallow();
      break;
    default:
      component = shallow(nodeWithProps);
      break;
  }
  return component;
}

/**
 *
 * @param node React Native Component
 * @param listOfProps List of Props that you pass into current Component
 * @param type Type of current Component ('single' or 'hoc')
 * @param styleName Name of current style (ex: 'style', 'titleStyle', 'containerStyle')
 */
export function getStyle<C extends Component, P = C['props'], S = C['state']>(
  node: ReactElement<P>,
  listOfProps: any,
  type: EnumComponent,
  styleName: 'style' | 'containerStyle' | 'iconStyle' | 'iconContainerStyle' | 'avatarStyle' | 'overlayContainerStyle' | 'placeholderStyle' | 'titleStyle' | 'badgeStyle' | 'textStyle' | 'buttonStyle' | 'buttonContainerStyle' | 'disabledStyle' | 'disabledTitleStyle' | 'disabledTextStyle' | 'disabledSelectedStyle' | 'disabledSelectedTextStyle' | 'innerBorderStyle'| 'selectedButtonStyle' | 'selectedTextStyle' | 'loadingStyle' | 'dividerStyle' | 'featuredSubtitleStyle' | 'featuredTitleStyle' | 'imageStyle' | 'imageWrapperStyle' | 'wrapperStyle' | 'backgroundImageStyle' | 'leftContainerStyle' | 'centerContainerStyle' | 'rightContainerStyle' | 'barStyle' | 'disabledInputStyle' | 'inputContainerStyle' | 'inputStyle' | 'labelStyle' | 'leftIconContainerStyle' | 'rightIconContainerStyle' | 'contentContainerStyle' | 'rightContentContainerStyle' | 'subtitleStyle' | 'rightTitleStyle' | 'rightSubtitleStyle' | 'overlayStyle' | 'thumbStyle' | 'trackStyle' | 'fontStyle' | 'h1Style' | 'h2Style' | 'h3Style' | 'h4Style' | 'captionStyle' | 'imageContainerStyle' = 'style',
) {
  /**
  * Use Mount (For HOCs Component) => THROW WARNING
  */
  // const wrapper = mount(<Text
  //   margin={10}
  //   marginVertical={10}
  //   marginHorizontal={10}
  //   marginTop={10}
  //   marginBottom={10}
  //   marginLeft={10}
  //   marginRight={10}
  // />);
  // const testStyle = wrapper.findWhere((node) => node.prop('testID') === 'EText')
  // .first().props().style;

  /**
  * Use Shallow (For HOCs Component) => OK
  */
  const component: any = getComponent(node, listOfProps, type);
  const testStyle = component.props()[styleName];
  return testStyle;
}

/**
 *
 * @param node React Native Component
 * @param listOfProps List of Props that you pass into current Component
 * @param type Type of current Component ('single' or 'hoc')
 * @param styleName Name of current style (ex: 'style', 'titleStyle', 'containerStyle')
 */
export function testStyleProps<C extends Component, P = C['props'], S = C['state']>(
  node: ReactElement<P>,
  listOfProps: any,
  type: EnumComponent,
  styleName: 'style' | 'containerStyle' | 'iconStyle' | 'iconContainerStyle' | 'avatarStyle' | 'overlayContainerStyle' | 'placeholderStyle' | 'titleStyle' | 'badgeStyle' | 'textStyle' | 'buttonStyle' | 'buttonContainerStyle' | 'disabledStyle' | 'disabledTitleStyle' | 'disabledTextStyle' | 'disabledSelectedStyle' | 'disabledSelectedTextStyle' | 'innerBorderStyle'| 'selectedButtonStyle' | 'selectedTextStyle' | 'loadingStyle' | 'dividerStyle' | 'featuredSubtitleStyle' | 'featuredTitleStyle' | 'imageStyle' | 'imageWrapperStyle' | 'wrapperStyle' | 'backgroundImageStyle' | 'leftContainerStyle' | 'centerContainerStyle' | 'rightContainerStyle' | 'barStyle' | 'disabledInputStyle' | 'inputContainerStyle' | 'inputStyle' | 'labelStyle' | 'leftIconContainerStyle' | 'rightIconContainerStyle' | 'contentContainerStyle' | 'rightContentContainerStyle' | 'subtitleStyle' | 'rightTitleStyle' | 'rightSubtitleStyle' | 'overlayStyle' | 'thumbStyle' | 'trackStyle' | 'fontStyle' | 'h1Style' | 'h2Style' | 'h3Style' | 'h4Style' | 'captionStyle' | 'imageContainerStyle' = 'style',
  level: number = 1,
) {
  const component: any = getComponent(node, listOfProps, type);
  let testStyle;
  switch (level) {
    case 1:
      testStyle = component.props()[styleName];
      break;
    case 2:
      testStyle = component.props().children.props[styleName];
      break;
    case 3:
      testStyle = component.props().children.props.children.props[styleName];
      break;
    default:
      testStyle = component.props()[styleName];
  }
  for (const [key, value] of Object.entries(testStyle)) {
    if (key in listOfProps) {
      expect(value).toBe(listOfProps[key]);
    }
  }
}

/**
 *
 * @param node React Native Component
 * @param type Type of current Component ('single' or 'hoc')
 * @param level Children level
 */
export function basicTest <C extends Component, P = C['props'], S = C['state']>(
  node: ReactElement<P>,
  type: EnumComponent,
  level: number = 1,
  haveChildren: boolean = true,
  haveStyleProp: boolean = true,
) {
  it('should render without issues', () => {
    const component: any = getComponent(node, {}, type);
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  if (haveChildren) {
    it('should have text as children', () => {
      const props: any = {
        children: 'Children Text',
      };
      const component: any = getComponent(node, props, type);
      switch (level) {
        case 1:
          expect(component.props().children).toBe('Children Text');
          break;
        case 2:
          expect(component.props().children.props.children).toBe('Children Text');
          break;
        case 3:
          expect(component.props().children.props.children.props.children).toBe('Children Text');
          break;
        default:
          expect(component.props().children).toBe('Children Text');
      }
    });
  }
  if (haveStyleProp) {
    it('should apply style prop', () => {
      const props = {
        style: { color: 'red', fontSize: 30 },
      };
      const testStyle = getStyle(node, props, type);
      const component: any = getComponent(node, props, type);
      switch (level) {
        case 1:
          expect(testStyle).toMatchObject(props.style);
          break;
        case 2:
          expect(component.props().children.props.style).toMatchObject(props.style);
          break;
        case 3:
          expect(component.props().children.props.children.props.style).toMatchObject(props.style);
          break;
        default:
          expect(testStyle).toMatchObject(props.style);
      }
    });
  }
}
