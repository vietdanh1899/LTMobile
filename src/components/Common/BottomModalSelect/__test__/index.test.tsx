import React from 'react';
import { EnumComponent, basicTest } from '@utils/testHelper';
import BottomModalSelect from '..';

const data = [
  { id: '0', name: 'Mở bán dự án Residences Quy Nhơn' },
  { id: '17', name: 'Công bố dự án Phúc Yên Prosper Phố Đông Thủ Đức' },
  { id: '3', name: 'Công bố dự án Century City Long Thành' },
  { id: '5', name: 'Mở bán dự án Green Dragon City Quảng Ninh' },
  { id: '6', name: 'Mở bán dự án Red Dragon City Hà Nội' },
  { id: '7', name: 'Mở bán dự án Violet Dragon City Hải Phòng' },
  { id: '8', name: 'Mở bán dự án Black Dragon City Lào Cai' },
];
const target = <BottomModalSelect data={data} />;
const type = EnumComponent.SINGLE;

describe('BottomModalSelect', () => {
  basicTest(target, type, 1, false, false);
});
