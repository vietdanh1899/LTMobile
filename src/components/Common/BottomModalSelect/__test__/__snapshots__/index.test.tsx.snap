// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomModalSelect should render without issues 1`] = `
<View
  testID="BottomModalSelect"
/>
`;
